zcash_client_backend/src/fees/common.rs

```diff
@@ -0,0 +1,122 @@
+use zcash_primitives::{
+    consensus::{self, BlockHeight},
+    memo::MemoBytes,
+    transaction::{
+        components::amount::{BalanceError, NonNegativeAmount},
+        fees::{transparent, FeeRule},
+    },
+};
+
+use super::{sapling, ChangeError, ChangeValue, DustAction, DustOutputPolicy, TransactionBalance};
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn single_change_output_balance<
+    P: consensus::Parameters,
+    NoteRefT: Clone,
+    F: FeeRule,
+    E,
+>(
+    params: &P,
+    fee_rule: &F,
+    target_height: BlockHeight,
+    transparent_inputs: &[impl transparent::InputView],
+    transparent_outputs: &[impl transparent::OutputView],
+    sapling_inputs: &[impl sapling::InputView<NoteRefT>],
+    sapling_outputs: &[impl sapling::OutputView],
+    dust_output_policy: &DustOutputPolicy,
+    default_dust_threshold: NonNegativeAmount,
+    change_memo: Option<MemoBytes>,
+) -> Result<TransactionBalance, ChangeError<E, NoteRefT>>
+where
+    E: From<F::Error> + From<BalanceError>,
+{
+    let overflow = || ChangeError::StrategyError(E::from(BalanceError::Overflow));
+    let underflow = || ChangeError::StrategyError(E::from(BalanceError::Underflow));
+
+    let t_in = transparent_inputs
+        .iter()
+        .map(|t_in| t_in.coin().value)
+        .sum::<Option<_>>()
+        .ok_or_else(overflow)?;
+    let t_out = transparent_outputs
+        .iter()
+        .map(|t_out| t_out.value())
+        .sum::<Option<_>>()
+        .ok_or_else(overflow)?;
+    let sapling_in = sapling_inputs
+        .iter()
+        .map(|s_in| s_in.value())
+        .sum::<Option<_>>()
+        .ok_or_else(overflow)?;
+    let sapling_out = sapling_outputs
+        .iter()
+        .map(|s_out| s_out.value())
+        .sum::<Option<_>>()
+        .ok_or_else(overflow)?;
+
+    let fee_amount = fee_rule
+        .fee_required(
+            params,
+            target_height,
+            transparent_inputs,
+            transparent_outputs,
+            sapling_inputs.len(),
+            if sapling_inputs.is_empty() {
+                sapling_outputs.len() + 1
+            } else {
+                std::cmp::max(sapling_outputs.len() + 1, 2)
+            },
```

Comment:
I presume this will change to use `sapling::builder::BundleType::num_outputs(sapling_inputs.len(), sapling_outputs.len() + 1)?`

Reply:
That's correct.